### PR TITLE
Use template rotation

### PR DESCRIPTION
**Pull request by tfussell**
According to Cluster API documentation, the recommended method for applying changes to clusters using ClusterClass is to use rotation (https://github.com/kubernetes-sigs/cluster-api/blob/72450d4d2a24e1f3c3737ffab6e599701b8a517f/docs/book/src/tasks/experimental-features/cluster-class/change-clusterclass.md#changing-clusterclass-templates). We therefore need to make this work with the helm chart approach.

To accomplish this, I have devised a method to calculate a hash of inputs and append it to template names. I noted where in the code this is happening below.

### CODE REVIEW

File: helm/cluster-openstack/templates/_helpers.tpl [hunk 1]
```diff
@@ -42,6 +41,50 @@ room for such suffix.
...
+  "imageName" (include "imageName" .)
+  "externalNetworkID" .Values.externalNetworkID
+  "rootVolume" .Values.rootVolume
+  "nodeClasses" .Values.nodeClasses
+  "controlPlane" (dict
+    "machineFlavor" .Values.controlPlane.machineFlavor
+    "diskSize" .Values.controlPlane.diskSize)
+  "oidc" .Values.oidc) }}
+{{- mustToJson $inputs | toString | quote | sha1sum | trunc 8 }}
+{{- end -}}
```
Comment (tfussell) [inline helm/cluster-openstack/templates/_helpers.tpl:58]: Here's the most important part. Generate a global hash of all input values that affect any template CRs. This hash can then be appended to all template names. Changing any input will lead to a new hash giving template rotation managed by helm.

**Comment by tfussell** [comment]
I don't plane to merge this in the current state, but I'm requesting a preliminary review from @giantswarm/team-rocket to get feedback on the general approach.

File: helm/cluster-openstack/templates/_cluster.tpl [hunk 2]
```diff
@@ -3,30 +3,33 @@ apiVersion: cluster.x-k8s.io/v1beta1
...
+          {{- include "labels.common" . | nindent 10 }}
     workers:
       machineDeployments:
+      {{- $outerScope := . }}
       {{- range .Values.nodePools }}
-      - class: {{ .class }}
-        name: {{ .name }}
-        replicas: {{ .replicas }}
+      {{- $innerScope := merge (deepCopy .) (deepCopy $outerScope) }}
+      {{- $nodeClass := merge (dict "name" .class) (deepCopy (get $outerScope.Values.nodeClasses .class)) (deepCopy $outerScope) }}
```
Comment (pipo02mix) [inline helm/cluster-openstack/templates/_cluster.tpl:26]: this is a bit complex to understand, what about adding a comment about why you do it?

**Review by pipo02mix — APPROVED** [review]
LGTM though I would like someone else check it

**Comment by tfussell** [comment]
Closing as outdated.

**Pull request closed without merging**